### PR TITLE
Fix handling of "Date array" type

### PR DESCRIPTION
**Pull request by bvenky27**
### Description
Array of type 'Date' doesn't work

#### Related issues

- connect to #1632 

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by jannyHou** [comment]
@slnode test please

**Comment by bvenky27** [comment]
@jannyHou I could see that the Commit Linter check was not successful. When I tried to check the details but clicking the 'Details' button, Its says 'HTTP ERROR 404 Not Found'. Can you please let me know what is the issue?

**Comment by dhmlau** [comment]
@bvenky27 , thanks for your PR.  The commit linter is run behind the firewall, that's why you cannot view the details.  Here is the error:
```
**************************************************
**
**  Linting commit logs
**
**  5 problems found:
**    f50885c - Fix bug with issue #1632: Array of type 'Date' does: First line should be 50 characters or less (saw 59)
**    1d12365 - Revert "Fixed bug with issue #1632: Array of type ': First line should be 50 characters or less (saw 69)
**    c780ec9 - Fixed bug with issue #1632: Array of type 'Date' no: First line should be 50 characters or less (saw 60)
**    fd8fc44 - Revert "Fixed bug with issue #1632: Array of type D: First line should be 50 characters or less (saw 67)
**    8c8880f - Fixed bug with issue #1632: Array of type Date not : First line should be 50 characters or less (saw 58)
**
**************************************************
```

As a reference, please see our commit message guideline for details: https://loopback.io/doc/en/contrib/git-commit-messages.html.  Thanks!

**Comment by bvenky27** [comment]
@dhmlau I see. So, the commit message is too big. Can you please suggest me on what to do to fix it? Do I need to revert the changes and commit it with new commit message with characters less than 50?

**Comment by bvenky27** [comment]
@dhmlau I pushed a commit with commit message less than 50 characters. Still its the same.

**Comment by dhmlau** [comment]
@bvenky27 , you'd need to amend the messages of the old commits, because the new commit won't replace the messages in the older commit. Looking at the commit messages, I think it's the best to squash your commits into one and edit the message at that time.  You can follow the instructions in here: https://github.com/wprig/wprig/wiki/How-to-squash-commits.

Please let me know if it works.  If not, @jannyHou or I can always help you to change the message.  :)

As a side note, you can change the commit message by `git commit --amend`.  More detailed instructions can be found here: https://help.github.com/articles/changing-a-commit-message/.

Hope it helps.

**Comment by bvenky27** [comment]
@dhmlau @jannyHou I did amend the old commit but still it failed. So, I used squash as suggested and made the commit now.

**Comment by bvenky27** [comment]
@dhmlau @jannyHou thanks for the assistance. I was able to fix the commit messages.

**Comment by bvenky27** [comment]
@jannyHou @dhmlau What is the status of review?

**Comment by jannyHou** [comment]
@bvenky27 I am not very export with the Date type in juggler, may take 1-2 days to review your PR. 
Sorry for the delay.

**Comment by bvenky27** [comment]
@jannyHou @dhmlau  can you please let me know the status of the review? Do I have to rebase the PR?

File: lib/list.js [hunk 1]
```diff
@@ -64,7 +64,9 @@ function List(items, itemType, parent) {
   }
 
   List.prototype.toItem = function(item) {
-    return isClass(this.itemType) ? new this.itemType(item) : this.itemType(item);
+    return (isClass(this.itemType) || (this.itemType.toString() == Date.toString())) ?
```
Comment (bajtos) [inline lib/list.js]: Let's extract this condition into a named variable to keep the code simpler to read.

```js
const isClassLike = isClass(this.itemType) || this.itemType.toString() == Date.toString();
return isClassLike ? new this.itemType(item) : this.itemType(item);
```

Also, do we need to convert `Date` constructor to string? Isn't it enough to compare the functions?

```js
const isClassLike = isClass(this.itemType) || this.itemType == Date;
```

**Comment by bajtos** [comment]
Closing in favor of #1702

**Comment by bvenky27** [comment]
Hey @bajtos, is the issue fixed without having to merge this changes?

**Comment by bajtos** [comment]
@bvenky27  Yes, it is! See the current implementation in `master`:

https://github.com/strongloop/loopback-datasource-juggler/blob/3fba79e725fe7752da19fb75fd0226a0896cb16d/lib/list.js#L67-L78

It handles few additional edge cases that were not handled by your patch, e.g. `null` value for `Date` properties.

The fix was backported to 3.x too, see  https://github.com/strongloop/loopback-datasource-juggler/pull/1719

**Comment by bvenky27** [comment]
@bajtos cool. That look good. Thank you.

**Pull request closed without merging**